add script for removing unused source tarballs in lookaside cache

dturecek opened the pull request:
The script goes through all branches for each project in lookaside cache and keeps directories mentioned in `sources` file, deleting all others.

clime commented:
Hey dturecek,

the script you provided is very deployment specific as it assumes there is three-levels deep hiearchy under `/var/lib/dist-git/cache/lookaside/pkgs/`. That is the case for copr-dist-git deployment but it is not the case for other deployments like Fedora DistGit.

I think this script would be a better fit to be placed in code for your particular deployment (e.g. copr-dist-git). The reason is the 'sources' file and its structure is defined client-side, not server-side and copr-dist-git uses the pyrpkg client library where the format is defined and hence you can provide a script that relies on that format in copr-dist-git package. Here, you would be basically just guessing the format.

However, if you want, we can place this script (as a non-executable) under `docs/scripts/dist-git` to serve as an example but it would require more work then.

praiskup commented:
> the script you provided is very deployment specific as it assumes there is three-levels deep hiearchy under /var/lib/dist-git/cache/lookaside/pkgs/. That is the case for copr-dist-git deployment but it is not the case for other deployments like Fedora DistGit.

Agreed, if anything was to be added to this code base, it should work against (a) concrete git repository, and (b) concrete lookaside cache sub-directory.

> the 'sources' file and its structure is defined client-side

It is, but there are known patterns which can be taken into account (I know about two formats).

praiskup commented:
@dturecek  I sort of like `pygit2` for these kinds of git repo operations (it would be probably safer, too).

dturecek commented:
I've changed the script so it now only goes over a single project and remove associated sources and is called like this: `./remove_old_sources project_name path_to_lookaside path_to_gitdir`. I'm adding a second script to copr that would then call the script here for each project in the lookaside cache.

What do you think about this?

clime commented:
> I've changed the script so it now only goes over a single project and remove associated sources and is called like this: `./remove_old_sources project_name path_to_lookaside path_to_gitdir`. I'm adding a second script to copr that would then call the script here for each project in the lookaside cache.
> 
> What do you think about this?

Well, the argument is still the same. The script uses and parses the sources file, whereas DistGit's scripts shouldn't care about the sources file.

xsuchy commented:
> Well, the argument is still the same. The script uses and parses the sources file, whereas DistGit's scripts shouldn't care about the sources file.

Why? It is part of the dist-git format. See README.md "...an additional text file called sources that contains link to the source tarball in the lookaside cache."

clime commented:
> Why? It is part of the dist-git format. See README.md "...an additional text file called sources that contains link to the source tarball in the lookaside cache."

Yup, that's all implemented client-side. Server side does not have any notion of sources file and its format for each individual repo.

xsuchy commented:
It does not have the sense to pretend that we do not know anything about the format - especially when this is just helper script - when 100 % of deployment put there sources and spec files. :)

clime commented:
> It does not have the sense to pretend that we do not know anything about the format - especially when this is just helper script - when 100 % of deployment put there sources and spec files. :)

Sry, idk what you exactly mean like this but there is no pretending. The current scripts know nothing about the sources format or even about the existence of the file. That's a client-side business, not server-side and it would be good not to introduce additional complexity by caring about the format on two different places. Please, implement what you need in copr-dist-git. We can include this script in the example directory as i proposed earlier.

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,62 @@
+    pushd "$cloned_dir" > /dev/null
+else
+    popd > /dev/null
+    exit 2
```
popd before exit doesn't make sense

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,62 @@
+
+eval "$(git for-each-ref --shell --format='branches+=(%(refname))' refs/remotes/)"
+for branch in "${branches[@]##*/}"; do
+    git checkout "$branch" &> /dev/null
```
You can use `git show REF:sources`, without the need to checkout ... doing checkout in git repository changes its state ...

praiskup replied:
So it allows you to avoid the `"${branches[@]##*/}"` brain twister :-)

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,62 @@
+    if [ -f sources ]; then
+        while read line; do
+	    ll=($line)
+	    whitelist+=(${ll[1]})
```
Note that sources line can contain either the old format:
`64c54d43787339a7cced48390eb3e1d0  uucp-1.07.tar.gz`
or the new one:
`SHA512 (tar-1.30.tar.xz.sig) = 149bfb112aae05e1f40a26ab2570fef71cf736eac62bf3a1bc8df905270c745500fc82764867323e440c67b1dd7116f5b0e1fd1977be7f10b333d481da3355e8`

Note that the name `ll` is not really obvious :-) and more common (and portable) statement is: 
```bash
set -- $line
filename=$2
```

praiskup replied:
plus `test $# -gt 2 && filename=$(echo "$filename" | sed -e "s/^(//" -e "s/)$//")`

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,62 @@
+pushd /tmp > /dev/null
+
+git clone "$pkg_git_dir" 2> /dev/null
+cloned_dir=`basename "$package"`
```
Ah, this is IMO not needed -- you can use the `git show REF:sources` in original repository.

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,62 @@
+
+package="$1"
+lookaside_dir="$2"
+git_dir="$3"
```
What about to have two arguments only, pkg_git_dir and pkg_lookaside_dir ...

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,62 @@
+    if [[ "${whitelist[@]}" =~ "${srcdir}" ]]; then
+        continue
+    else
+    	rm -rf "$source_dir"
```
I'd maybe do only `rm "$source_dir"`, at least at the beginning it is slightly safer variant ...

praiskup commented:
The "helper" script is not only related to copr ... and it might be convenient to anyone.  So it would be really pity if that lived only in copr-dist-git.  The only think is that the script should self-document itself, what it does, etc....  users need to be sure that all tarballs are removed, except for the files referenced by **latest versions of branches** (e.g, some users might prefer to keep at least a few commits of git history rebuildable, and thus this is not for them (yet)).

dturecek commented:
I've addressed your points except for the `rm (-rf) "$source_dir"` as the tarballs are hidden inside the directory of the same name, for example `copr-frontend-1.137.tar.gz/8cade45c687ce72c6649e0097b93b0ae/copr-frontend-1.137.tar.gz`

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,56 @@
+    if [[ "${whitelist[@]}" =~ "${srcdir}" ]]; then
+        continue
+    else
+        rm -rf $source_dir
```
missing double quotes

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,56 @@
+        filename=$2
+        test $# -gt 2 && filename=$(echo "$filename" | sed -e "s/^(//" -e "s/)$//")
+        whitelist+=($filename)
+    done < <(git show $branch:sources)
```
I'd prefer double quoting $filename and $branch as well.

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,56 @@
+    exit 1
+fi
+
+pushd $pkg_git_dir > /dev/null
```
Double quotes.

praiskup commented:
How do you plan to install the script?  Should that be in the RPM?

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,56 @@
+for source_dir in "$pkg_lookaside_dir"/*; do
+    srcdir="${source_dir##*/}"
+
+    if [[ "${whitelist[@]}" =~ "${srcdir}" ]]; then
```
I have to admit this was new for me, but it doesn't seem to work as you expect?:
```
$ [[ "${a[@]}" =~ b ]] && echo yes
yes
```
I'd rather iterate over the whitelist: `for dir in "${whitelist[@]}"; do test "$dir" = "$srcdir"; ... ; done`.

dturecek commented:
I've added the missing double quotes, changed the way of checking the whitelist, and added the symlink creation to the spec file so that the script is installed the same way as the other three scripts there.

praiskup commented on dist-git.spec:
```diff
@@ -331,5 +332,3 @@ fi
 * Mon Mar 30 2015 Adam Samalik <asamalik@redhat.com> 0.2-1
 - new package built with tito
-
-
```
Whitespace fixes are for different PR.

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,66 @@
+
+    for dir in "${whitelist[@]}"; do
+        test "$dir" = "$srcdir";
+        if [[ $? = 0 ]]; then
```
`if test "$dir" = "$srcdir"; then` ?

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,66 @@
+        fi
+    done
+
+    if [[ "$keep" = true ]]; then
```
if $keep; then

praiskup replied:
or even `$keep && continue`

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,66 @@
+        test "$dir" = "$srcdir";
+        if [[ $? = 0 ]]; then
+            keep=true
+            continue
```
You probably wanted `break` here.

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/bash
```
#! /bin/bash is better

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,65 @@
+popd > /dev/null
+
+# remove all sources that are not referenced
+for source_dir in "$pkg_lookaside_dir"/*; do
```
This could work:
```bash
#! /bin/bash

pkg_lookaside_dir=/var/lib/dist-git/cache/lookaside/pkgs/praiskup/ping-temp/tar

# Read the old format directory (filename/md5sum/filenaem)
while read file; do
    old_IFS=$IFS
    IFS=/
    set -- $file
    IFS=$old_IFS
    # safety measure, if this is really the layout we expect the first and
    # third part matches
    test "$1" = "$3" || {
        echo "skipped $file"
        continue
    }
    echo unlink "$pkg_lookaside_dir/$file"
done < <( cd "$pkg_lookaside_dir" ; find -mindepth 3 -maxdepth 3 -type f  -printf '%P\n' )
```

Once anyone feels like the "new" format needs to be supported, it can easily be added.

praiskup replied:
I added /var/tmp/test-crawl-script on stg dist-git for testing:
```
# /var/tmp/test-crawl-script /var/lib/dist-git/cache/lookaside/pkgs/praiskup/ping-temp/tarskipped tar-1.32.tar.xz/83e38700a80a26e30b2df054e69956e5/tar-1.32.tar.xz.DONT_DELETE
unlink /var/lib/dist-git/cache/lookaside/pkgs/praiskup/ping-temp/tar/tar-1.32.tar.xz/83e38700a80a26e30b2df054e69956e5/tar-1.32.tar.xz
```

dturecek commented:
I've changed the script based on your suggestion. In addition, I now check if the hash in the directory name matches the one in the `source` files. 
If the source file uses the new format then the whole project is skipped (and the script ends).

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,75 @@
+        hash=$1
+        filename=$2
+        # skip projects using the new format
+        test $# = 2 -a $filename = `echo "$filename" | sed -e "s/^(//" -e "s/)$//"` || exit 0
```
This works, though sligthly nicer would be:
```bash
test $# -eq 0 && continue # empty line
test $# -ne 2 && exit 1   # we parse only the old <SUM><SPACE><FILENAME>format
```
Probably some error message saying why we `exit 1` would be nice as well, e.g.
through `die()` function.

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,75 @@
+whitelist=()
+
+# find sources that are referenced by the latest commit in any of the branches
+for branch in `git for-each-ref --format="%(refname:short)" refs/heads`; do
```
shellcheck: Use $(...) notation instead of legacy backticked `...`. [SC2006]

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,75 @@
+
+# find sources that are referenced by the latest commit in any of the branches
+for branch in `git for-each-ref --format="%(refname:short)" refs/heads`; do
+    while read line; do
```
shellcheck: read without -r will mangle backslashes. [SC2162]

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,75 @@
+    exit 1
+fi
+
+pushd "$pkg_git_dir" > /dev/null
```
shellcheck: Use 'cd ... || exit' or 'cd ... || return' in case cd fails. [SC2164]
(use pushd || exit 1)

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,75 @@
+    IFS=$old_IFS
+    "$keep" && continue
+    unlink "$pkg_lookaside_dir/$file"
+done < <( cd "$pkg_lookaside_dir" ; find -mindepth 3 -maxdepth 3 -type f  -printf '%P\n' )
```
`cd "$pkg_lookaside_dir" || exit 1`

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,75 @@
+    keep=false
+    for source in "${whitelist[@]}"; do
+        IFS=','
+        read -ra parsedSource <<< "$source"
```
I'd still prefer `set -- $source` here.  It would override your `$1` and `$2`, but it's even better to do
```bash
filename=$1
hash=$2
```
... above, and do again whatever we want with `$@`.

praiskup commented:
This starts to look really nice, thanks @dturecek!

dturecek commented:
I've addressed your points, please take a look.

praiskup commented:
+1, this looks nice to me

praiskup commented on scripts/dist-git/remove_unused_sources:
```diff
@@ -0,0 +1,77 @@
+    exit 1
+fi
+
+pushd "$pkg_git_dir" > /dev/null || exit
```
s/exit/exit 1/ recomended

praiskup commented:
+1, LGTM

xsuchy commented:
Nice work.
